### PR TITLE
DAOS-5703 dfuse: Add mknod call and use special files for UNS.

### PR DESCRIPTION
**Pull request by ashleypittman**
Rather than using directories for UNS entry point use the
mode bits for FIFO.
Change dfs to accept FIFO as a mode, and to treat them as
directories, but not allow readdir or creating files.
Change dfuse to allow creation of FIFOs, but once they're
created then report them as dirs.

Resolve a bug in NLT where cleanup of agent dirs was happening
early, leading to errors on failure paths.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

**Review by daltonbohning — COMMENTED** [review]
Something with the `mknod` implementation is not working, but I'm not sure why.
Test case:
```
test_file = /path/to/dfuse/test_file
mknod(test_file, S_IFREG | S_IRWXO | S_IRWXU | S_IRWXG, dev);
// success
[dbohninx@boro-43 dfuse]$ ls -l test_file
-rwxrwxr-x 1 dbohninx dbohninx 0 Nov 30 18:02 test_file
// okay, good
[dbohninx@boro-43 dfuse]$ chmod -x test_file
chmod: changing permissions of ‘test_file’: Operation not permitted
// doesn't work, but it should
```
Whereas, if I create a file with `touch`, the same case works:
```
[dbohninx@boro-43 dfuse]$ touch file
[dbohninx@boro-43 dfuse]$ chmod 775 file
[dbohninx@boro-43 dfuse]$ ls -l file
-rwxrwxr-x 1 dbohninx dbohninx 0 Nov 30 18:07 file
[dbohninx@boro-43 dfuse]$ chmod -x file
[dbohninx@boro-43 dfuse]$ ls -l file
-rw-rw-r-- 1 dbohninx dbohninx 0 Nov 30 18:07 file
// good
```

Thoughts?

**Comment by daltonbohning** [comment]
I did some more testing of the permissions issue I noted before. It doesn't seem to be just the mknod implementation that isn't working:
```
open("./dfuse/test_file", O_CREAT, S_IFREG | S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH);
// chmod ./dfuse/test_file fails with "Operation not permitted"
```

By default, the file is opened with O_RDONLY. If I open the file with O_WRONLY:
```
open("./dfuse/test_file", O_WRONLY | O_CREAT, S_IFREG | S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH);
// chmod success
```

So something is weird when creating the file in read-only mode.

**Comment by ashleypittman** [comment]
> I did some more testing of the permissions issue I noted before. It doesn't seem to be just the mknod implementation that isn't working:
> 
> ```
> open("./dfuse/test_file", O_CREAT, S_IFREG | S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH);
> // chmod ./dfuse/test_file fails with "Operation not permitted"
> ```
> 
> By default, the file is opened with O_RDONLY. If I open the file with O_WRONLY:
> 
> ```
> open("./dfuse/test_file", O_WRONLY | O_CREAT, S_IFREG | S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH);
> // chmod success
> ```
> 
> So something is weird when creating the file in read-only mode.

Can you make a ticket about that please?  I suspect it'll require changes to dfs and dfuse to resolve.

**Comment by daltonbohning** [comment]
> Can you make a ticket about that please? I suspect it'll require changes to dfs and dfuse to resolve

Sure, created: [DAOS-6238](https://jira.hpdd.intel.com/browse/DAOS-6238)

File: src/client/dfs/dfs.c
Comment (daltonbohning) [reply src/client/dfs/dfs.c:855]: Thanks!

**Review by daltonbohning — DISMISSED** [review]
I left a few comments, but otherwise this seems to be working as expected.
I did run some functional tests using mpifileutils/dcp, which utilizes both DFS and DFUSE read, write, mkdir, mknod, and UNS paths in a dfuse container.
The only issue is [DAOS-6238](https://jira.hpdd.intel.com/browse/DAOS-6238).
After putting in a workaround locally (adding O_RDWR to the dfs_open call for mknod), the functional tests at least seem to be in order.

File: src/client/dfs/dfs.c
Comment (ashleypittman) [reply src/client/dfs/dfs.c:3088]: Yes, this is a permissions check that should have been removed before.  The osetattr() one came up in testing, and this chmod() one is doing the same check, so I removed it as well.

File: src/client/dfuse/ops/mknod.c
Comment (ashleypittman) [reply src/client/dfuse/ops/mknod.c:33]: Not sure what happened here.  I'll fix.

File: src/client/dfuse/ops/mknod.c
Comment (ashleypittman) [reply src/client/dfuse/ops/mknod.c:47]: we probably should, I'll respin.  I removed O_DRWR as it doesn't seem possible to do any I/O on a file until it's been opened but it might interact badly with DAOS-6238 as you say.

I hope to have a PR for that issue tomorrow so maybe we could land that first and re-test this once that's landed.

Comment (daltonbohning) [reply src/client/dfuse/ops/mknod.c:47]: Sounds good to me. I do think this _**should**_ work without O_RDWR, so DAOS-6238 would give a more proper fix.

**Review by daltonbohning — APPROVED** [review]
This PR, along with #4166, seems to solve the "no permission" issue.
I ran manual tests for open, creat, mkdir, symlink, and mknod (this branch merged with #4166), and chmod worked for each case.
Also, for what it's worth, I ran some current and in progress datamover tests that run over dfuse files, dirs, symlinks, and UNS entries. No errors were reported.

File: src/client/dfs/dfs.c
Comment (mchaarawi) [inline src/client/dfs/dfs.c:855]: im not sure i understand this change. I don't understand the comment either.
what if dir->mode is not set? like in dfs_mount

File: src/client/dfs/duns.c
Comment (mchaarawi) [inline src/client/dfs/duns.c:655]: why do you resolve checkpatch errors? i think we should abide by the 80 char limit.

File: src/client/dfuse/dfuse.h
Comment (mchaarawi) [inline src/client/dfuse/dfuse.h:133]: im not familiar with fuse callbacks TBH, but why is this a replacement for mkdir?

File: src/client/dfuse/ops/readdir.c
Comment (mchaarawi) [inline src/client/dfuse/ops/readdir.c:66]: so all that means that we will never support fifo over dfuse, right?

File: src/client/dfs/dfs.c
Comment (ashleypittman) [reply src/client/dfs/dfs.c:855]: This function is called for both dirs and fifos, so the check is verifying that if we expect it to be a directory then it is a directory.  It should possibly be checking that dir->mode I S_IFMT = entry.mode | S_IFMT.  Essentially that the directory hasn't been replaced with a entry of a different type since it was opened.

File: src/client/dfuse/dfuse.h
Comment (ashleypittman) [reply src/client/dfuse/dfuse.h:133]: fuse will call either, mknod is just a more generic version of mkdir.  There are small changes in the dfuse_cont file to check if the correct file type is specified there.

File: src/client/dfuse/ops/readdir.c
Comment (ashleypittman) [reply src/client/dfuse/ops/readdir.c:66]: Yes, but we never had any plans to do that.

File: src/client/dfs/dfs.c
Comment (mchaarawi) [reply src/client/dfs/dfs.c:855]: well as i mentioned what if we are not setting what we expect?  we only do that in dfs_open.
and to make this clear, maybe add explicit checks here for fifo and dir.

Comment (ashleypittman) [reply src/client/dfs/dfs.c:855]: Looking at this closer I'm not sure the check is needed.  open_dir is either called in dfs_mount on (/) which will always be a directory, or within a transaction.  There might be a race where one client tries to make a directory, but another client makes a file, but the transaction should catch this so maybe this check could just be removed entirely.

The check was here before anyway, I just extended it to not perform the check if the entry in question is not a directory (and therefore is a fifo) but perhaps the check simply isn't needed.

I don't know what would happen if you called dfs_open(S_IFDIR) but without O_CREATE and a file existed at that name, this check would catch that but that's an odd case that's probably not allowed in POSIX anyway.  DFuse will always pass O_CREAT to dfs_open(), for "open" it uses dfs_dup().

File: src/client/dfuse/ops/readdir.c
Comment (mchaarawi) [reply src/client/dfuse/ops/readdir.c:66]: well my question was what if someone is interested in having this in the future?
do we want to close that door?

Comment (ashleypittman) [reply src/client/dfuse/ops/readdir.c:66]: We've discussed this in the POSIX calls and with the customer several times, nobody has ever expressed an indication that we might want to.

There are several other bits we could use, block device or character files etc that would probably work just as well assuming fuse supports them if you think that would be better.

File: src/client/dfs/dfs.c
Comment (mchaarawi) [reply src/client/dfs/dfs.c:855]: well we do want to make sure that the entry is of the correct type?
what if one calls opendir on a file? we can't have this succeed.. not sure how that is not allowed. it can be called directly from the DFS API.

Comment (ashleypittman) [reply src/client/dfs/dfs.c:855]: This is where we check for calling open_dir() on a file.  I've added S_IFDIR in the call from dfs_mount() and added a seperate check for FIFOs.  open_file() already has a check, and open_symlink doesn't support dfs_open() without O_CREATE.

File: src/client/dfuse/ops/readdir.c
Comment (mchaarawi) [reply src/client/dfuse/ops/readdir.c:66]: I'm just thinking beyond our current plan and how tied in we would be to this. I guess we could  change dfuse in that case to use something else.
but what i don't like is one could call dfs_open with FIFO and expect a FIFO, but they are not  getting that.. using the dfs API that is. what we are doing here is something specific / handled  only in dfuse
we definitely discussed implementing a block device in the future, so we can't use that..

Comment (mchaarawi) [reply src/client/dfuse/ops/readdir.c:66]: im ok with the patch. i just want to make sure that we understand what this exposes to the user and document this properly, that FIFO is not supported from DFS and it's reserved only for dfuse use really for now.
@johannlombardi @bfaccini is this OK to do?
i know lustre has the lmva that doesn't require this special case, but we don't have that in dfs and would add overhead in dir open if we do that.

**Review by mchaarawi — DISMISSED** [review]

File: src/client/dfuse/dfuse.h
Comment (ashleypittman) [reply src/client/dfuse/dfuse.h:338]: I've got a fix for this as well which I can include next time but I'll wait for another review to push anything.

**Comment by ashleypittman** [comment]
That last push is just to resolve the two checkpatch issues.

**Review by daltonbohning — APPROVED** [review]

**Review by mchaarawi — APPROVED** [review]

File: src/client/dfuse/ops/readdir.c
Comment (johannlombardi) [reply src/client/dfuse/ops/readdir.c:66]: We might eventually have to support VM mounting a DAOS container a root filesystem, so i think that dev/char/fifo files will eventually be required. Another approach would be to have on our own extra flags in DFS to represent links to external container. This could be in addition to IFDIR so maintain compatibility. We could also consider a "lookup + fetch EA" call to do it in a single RPC.
In any case, let's land this patch in the meantime and work on an improved version.

Comment (bfaccini) [reply src/client/dfuse/ops/readdir.c:66]: I also think that tp deflect inode's mode bits from their original purpose can lead to problems at some point.

**Pull request merged**
